genesyscloud/resource_genesyscloud_knowledge_knowledgebase.go

```diff
@@ -0,0 +1,202 @@
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/mypurecloud/terraform-provider-genesyscloud/genesyscloud/consistency_checker"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/mypurecloud/platform-client-sdk-go/v80/platformclientv2"
+)
+
+func getAllKnowledgeKnowledgebases(_ context.Context, clientConfig *platformclientv2.Configuration) (ResourceIDMetaMap, diag.Diagnostics) {
+	resources := make(ResourceIDMetaMap)
+	knowledgeAPI := platformclientv2.NewKnowledgeApiWithConfig(clientConfig)
+
+	for pageNum := 1; ; pageNum++ {
+		const pageSize = 100
+		knowledgeBases, _, getErr := knowledgeAPI.GetKnowledgeKnowledgebases("", "", "", fmt.Sprintf("%v", pageSize), "", "", false, "", "")
+		if getErr != nil {
+			return nil, diag.Errorf("Failed to get page of knowledge bases: %v", getErr)
+		}
+
+		if knowledgeBases.Entities == nil || len(*knowledgeBases.Entities) == 0 {
+			break
+		}
+
+		for _, knowledgeBase := range *knowledgeBases.Entities {
+			resources[*knowledgeBase.Id] = &ResourceMeta{Name: *knowledgeBase.Name}
+		}
+	}
+
+	return resources, nil
+}
+
+func knowledgeKnowledgebaseExporter() *ResourceExporter {
+	return &ResourceExporter{
+		GetResourcesFunc: getAllWithPooledClient(getAllKnowledgeKnowledgebases),
+		RefAttrs:         map[string]*RefAttrSettings{}, // No references
+	}
+}
+
+func resourceKnowledgeKnowledgebase() *schema.Resource {
+	return &schema.Resource{
+		Description: "Genesys Cloud Knowledge Base",
+
+		CreateContext: createWithPooledClient(createKnowledgeKnowledgebase),
+		ReadContext:   readWithPooledClient(readKnowledgeKnowledgebase),
+		UpdateContext: updateWithPooledClient(updateKnowledgeKnowledgebase),
+		DeleteContext: deleteWithPooledClient(deleteKnowledgeKnowledgebase),
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		SchemaVersion: 1,
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Description: "Knowledge base name",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"description": {
+				Description: "Knowledge base description",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"core_language": {
+				Description:  "Core language for knowledge base in which initial content must be created, language codes [en-US, en-UK, en-AU, de-DE] are supported currently, however the new DX knowledge will support all these language codes",
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringInSlice([]string{"en-US", "en-UK", "en-AU", "de-DE", "es-US", "es-ES", "fr-FR", "pt-BR", "nl-NL", "it-IT", "fr-CA"}, false),
+			},
+			"published": {
+				Description: "Flag that indicates the knowledge base is published",
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
```

Comment:
I got a plan not empty error when running the tests for this resource and I think it's because this field should be computed

Reply:
That's odd, the test for knowledgebase passes for me.  I'm wondering if it still fails if you do a git pull from this branch?

Reply:
Actually another comment is saying this test needs an update step, if you're running a different test for the resource that's probably why.  Will update

Reply:
Actually it passed for me on this attempt. I noticed when I create a knowledge base for the org we use in this repo's workflow, `published` comes back as true in the response, but with hollywoo it comes back false. Not sure if that means it should be computed or not. @ronanwatkins any thoughts?

Reply:
Yeah probably safe to put it as computed. There generally isn't any negative consequences of marking an attribute as computed.

> Actually another comment is saying this test needs an update step, that's probably why I'm not seeing the failure. Update steps are just good to have in the tests because it provides for a more comprehensive test